### PR TITLE
Support multi secret

### PR DESCRIPTION
**Pull request by hechao-ustc**
issue: https://github.com/asf-transfer/doris-operator/issues/237
Support multi secret:
```
spec:
  feSpec:
    secrets:
    # use kubectl create secret generic db-user --from-file=./username.txt -n doris
    - secretName: db-user
      mountPath: /etc/doris
      readOnly: true
    # use kubectl create secret generic db-pass --from-file=./password.txt -n doris
    - secretName: db-pass
      mountPath: /opt/doris
      readOnly: true
```

### CODE REVIEW

File: pkg/common/utils/k8s/client.go [hunk 1]
```diff
@@ -229,6 +229,20 @@ func GetConfigMap(ctx context.Context, k8scient client.Client, namespace, name s
 	return &configMap, nil
 }
 
+// CheckSecretExist check if the secret is in the namespace.
+func CheckSecretExist(ctx context.Context, k8scient client.Client, namespace string, secrets []dorisv1.Secret) {
```
Comment (catpineapple) [inline pkg/common/utils/k8s/client.go]: This func is not necessary. After checking, it only prints the log and has no effect on subsequent operations. Checking whether it exists can be done later.

File: pkg/controller/sub_controller/sub_controller.go [hunk 2]
```diff
@@ -125,6 +125,9 @@ func (d *SubDefaultController) CheckConfigMountPath(dcr *dorisv1.DorisCluster, c
 	var mountsMap = make(map[string]dorisv1.MountConfigMapInfo)
 	for _, cm := range cms {
 		path := cm.MountPath
+		if cm.MountPath == "" {
```
Comment (catpineapple) [inline pkg/controller/sub_controller/sub_controller.go]: Empty strings are also expected behavior and should not be treated specially.

Comment (hechao-ustc) [reply pkg/controller/sub_controller/sub_controller.go]: Ok, I'll modify this.

File: api/doris/v1/types.go [hunk 3]
```diff
@@ -281,6 +285,20 @@ type MountConfigMapInfo struct {
...
+	SecretName string `json:"secretName,omitempty"`
+
+	// Current Secret Mount Path, default is "/etc/doris"
+	// +optional
+	MountPath string `json:"mountPath,omitempty"`
+
+	// Mounted read-only if true, read-write otherwise (false or unspecified).
+	// Defaults to false.
+	// +optional
+	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,2,opt,name=readOnly"`
```
Comment (catpineapple) [inline api/doris/v1/types.go]: 1. There is no need to provide readonly configuration here, its default policy is false, we need this configuration to allow users to modify it midway
2. This attribute of protobuf does not need to be added

File: pkg/common/utils/k8s/client.go (hunk 1 above)
Comment (hechao-ustc) [reply pkg/common/utils/k8s/client.go]: I moved `CheckSecretExist` to after `k8s.ApplyStatefulSet`. I think this check is necessary. Without this check and without created the secret, the operator log will not contain the following error log, and the pod will always be in the `ContainerCreating` state.
image

**Review by catpineapple — APPROVED** [review]
LGTM

File: api/doris/v1/types.go [hunk 4]
```diff
@@ -281,6 +285,16 @@ type MountConfigMapInfo struct {
 	MountPath string `json:"mountPath,omitempty"`
 }
 
+type Secret struct {
+	// name of secret that needs to mount.
+	SecretName string `json:"secretName,omitempty"`
+
+	// Current Secret Mount Path, default is "/etc/doris"
+	// If Secret belongs to the same Secrets, their MountPath cannot be repeated.
+	// +optional
```
Comment (intelligentfu8) [inline api/doris/v1/types.go:304]: mountPath can't be repeated.

File: doc/examples/doriscluster-sample-secret.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# this yaml describe `secret` config in DorisCluster CRD.
+#Compareed to `doriscluster-sample.yaml`, It have `secret` config for fe
+apiVersion: doris.selectdb.com/v1
```
Comment (intelligentfu8) [inline doc/examples/doriscluster-sample-secret.yaml:36]: please add secrets that used in the example, in front of the yaml.

File: pkg/common/utils/k8s/client.go [hunk 6]
```diff
@@ -229,6 +229,20 @@ func GetConfigMap(ctx context.Context, k8scient client.Client, namespace, name s
 	return &configMap, nil
 }
 
+// CheckSecretExist check if the secret is in the namespace.
```
Comment (intelligentfu8) [inline pkg/common/utils/k8s/client.go]: ChecSecretExist, check the secret exist or not in specify namespace.

File: pkg/common/utils/resource/pod.go [hunk 7]
```diff
@@ -678,6 +688,50 @@ func getMultiConfigVolumeAndVolumeMount(cmInfo *v1.ConfigMapInfo, componentType
...
+	var volumeMounts []corev1.VolumeMount
+
+	if bSpec.Secrets == nil {
+		return volumes, volumeMounts
+	}
+
+	defaultMountPath := ""
+	switch componentType {
+	case v1.Component_FE, v1.Component_BE, v1.Component_CN, v1.Component_Broker:
+		defaultMountPath = config_env_path
```
Comment (intelligentfu8) [inline pkg/common/utils/resource/pod.go]: can't use env_config_path,should use secret_config_path

File: pkg/common/utils/resource/pod.go [hunk 8]
```diff
@@ -678,6 +688,50 @@ func getMultiConfigVolumeAndVolumeMount(cmInfo *v1.ConfigMapInfo, componentType
 	return volumes, volumeMounts
 }
 
+func getMultiSecretVolumeAndVolumeMount(bSpec *v1.BaseSpec, componentType v1.ComponentType) ([]corev1.Volume, []corev1.VolumeMount) {
+	var volumes []corev1.Volume
+	var volumeMounts []corev1.VolumeMount
+
+	if bSpec.Secrets == nil {
```
Comment (intelligentfu8) [inline pkg/common/utils/resource/pod.go]: this check is not needed. for golang splice, iterate the nil splice is ok.

File: pkg/controller/sub_controller/be/controller.go [hunk 9]
```diff
@@ -99,6 +100,11 @@ func (be *Controller) Sync(ctx context.Context, dcr *v1.DorisCluster) error {
 		return err
 	}
 
+	// check if the secret exists in the namespace
+	if dcr.Spec.BeSpec.Secrets != nil {
```
Comment (intelligentfu8) [inline pkg/controller/sub_controller/be/controller.go]: this check should before "applystatefulset" and should output an event on dorislcuster.

File: api/doris/v1/types.go (hunk 4 above)
Comment (hechao-ustc) [reply api/doris/v1/types.go:304]: Done

File: doc/examples/doriscluster-sample-secret.yaml (hunk 5 above)
Comment (hechao-ustc) [reply doc/examples/doriscluster-sample-secret.yaml:36]: Done

File: pkg/common/utils/k8s/client.go (hunk 6 above)
Comment (hechao-ustc) [reply pkg/common/utils/k8s/client.go]: Done, and I moved this method to sub_controller.go

File: pkg/common/utils/resource/pod.go (hunk 7 above)
Comment (hechao-ustc) [reply pkg/common/utils/resource/pod.go]: Done

File: pkg/common/utils/resource/pod.go (hunk 8 above)
Comment (hechao-ustc) [reply pkg/common/utils/resource/pod.go]: Done

File: pkg/controller/sub_controller/be/controller.go (hunk 9 above)
Comment (hechao-ustc) [reply pkg/controller/sub_controller/be/controller.go]: Done

**Review by intelligentfu8 — APPROVED** [review]

**Pull request merged**